tools/serverpod_cli/lib/src/analyzer/entities/stateful_analyzer.dart

```diff
@@ -0,0 +1,144 @@
+import 'package:serverpod_cli/analyzer.dart';
+import 'package:serverpod_cli/src/generator/code_generation_collector.dart';
+import 'package:serverpod_cli/src/util/protocol_helper.dart';
+
+class StatefulAnalyzer {
+  final Map<String, SerializableEntityAnalyzer> _analyzers = {};
+  List<SerializableEntityDefinition> _entities = [];
+
+  Function(Uri, CodeGenerationCollector)? _onErrorsChangedNotifier;
+
+  /// Validates all protocols by running the validator twice to make sure all
+  /// references are resolved. The state is preserved to make future validations
+  /// less expensive. This method is required to use for the initialization of
+  /// the state. Subsequent validations should use [validateAll] or [validateProtocol].
+  List<SerializableEntityDefinition> initialValidation(
+      List<ProtocolSource> sources) {
+    for (var yamlSource in sources) {
+      var analyzer = SerializableEntityAnalyzer(
+        yaml: yamlSource.yaml,
+        sourceFileName: yamlSource.uri.path,
+        subDirectoryParts: yamlSource.protocolRootPathParts,
+        collector: CodeGenerationCollector(),
+      );
+
+      _analyzers[yamlSource.uri.path] = analyzer;
+    }
+
+    _entities = _validateAllAnalyzers();
+    _entities = _validateAllAnalyzers(_entities);
+    return _entities;
+  }
+
+  /// Runs the validation on all protocols, assumes [initialValidation] has been
+  /// run before, if not this returns an empty list.
+  /// Errors are reported through the [onErrorsChangedNotifier].
+  List<SerializableEntityDefinition> validateAll() {
+    _entities = _validateAllAnalyzers(_entities);
+    return _entities;
+  }
+
+  /// Runs the validation on a single protocol, assumes [initialValidation] has been
+  /// run before. The protocol must exist in the state, if not this returns
+  /// the last validated state. To validate a new protocol, use [addYamlProtocol]
+  /// and then run [validateAll] or [validateProtocol].
+  /// Errors are reported through the [onErrorsChangedNotifier].
```

Comment:
This should probably be referring to the method as the actual property is private.